### PR TITLE
Build native images using GraalVM

### LINKED ISSUE #1123: Manage graalvm native images
Hi,

Is there any plan to support builds of native executables using graalvm. ie `native-image`.

if not, would you be interested in a prototype + a Pull Request ? I'm playing with it.

Thanks.

### PR DESCRIPTION
**Pull request by ScalaWilliam**
As per #1123.

This allows us to use GraalVM's ecosystem and create native binaries straight from SBT. High performance native binaries, native interop with many programming languages (JavaScript, Python, Ruby, R), .... Especially R and Python, super important for data science.

### CODE REVIEW

**Comment by ScalaWilliam** [comment]
I'd like feedback on the naming: the plugin name in particular - is it better for it to be GraalVM, or GraalVMNativeImage, or any other alternative?

**Comment by ScalaWilliam** [comment]
I thought about that @dborisenko as well. Could do this in a separate PR as it may need some abstraction of the Docker plugin.

File: src/main/scala/com/typesafe/sbt/packager/graalvm/Keys.scala [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+package packager
+package graalvm
+
+import sbt._
+
+/**
+  * GraalVM settings
+  */
+trait GraalVMKeys {
+  val nativeImageOptions = SettingKey[Seq[String]]("native-image-options", "native-image options")
```
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/graalvm/Keys.scala]: This is a graal only setting, right? For package specific settings we simply prefix them with the package name. In this case `graalVmNativeImageOptions` and don't scope them. This way they are clearly marked as format specific settings.

File: src/main/scala/com/typesafe/sbt/packager/graalvm/GraalVMPlugin.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  import autoImport._
+
+  override def projectConfigurations: Seq[Configuration] = Seq(GraalVM)
+
+  override lazy val projectSettings = Seq(
+    target in GraalVM := target.value / "graalvm",
+    sourceDirectory in GraalVM := sourceDirectory.value,
+    name in GraalVM := name.value,
```
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/graalvm/GraalVMPlugin.scala]: The name setting is not normalized. Are there any restrictions for the image name?

File: src/main/scala/com/typesafe/sbt/packager/graalvm/GraalVMPlugin.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    val GraalVM: Configuration = config("graalvm")
+  }
+
+  import autoImport._
+
+  override def projectConfigurations: Seq[Configuration] = Seq(GraalVM)
+
+  override lazy val projectSettings = Seq(
+    target in GraalVM := target.value / "graalvm",
+    sourceDirectory in GraalVM := sourceDirectory.value,
```
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/graalvm/GraalVMPlugin.scala]: Why is the source directory required?

File: src/main/scala/com/typesafe/sbt/packager/graalvm/GraalVMPlugin.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    name in GraalVM := name.value,
+    nativeImageOptions in GraalVM := Seq(s"-H:Name=${(name in GraalVM).value}"),
+    packageBin in GraalVM := {
+      (target in GraalVM).value.mkdirs()
+      val className = (mainClass in Compile).value.getOrElse(sys.error("Could not find a main class."))
+      val classpath = Seq((packageBin in Compile).value) ++ (dependencyClasspath in Compile).value.map(_.data)
```
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/graalvm/GraalVMPlugin.scala]: Depending on packageBin and dependencyClasspath in compile is probably fine. If users want to build graal images in test scope we'll make this more flexible.

**Comment by ScalaWilliam** [comment]
@muuki88 great review, thank you!

For Travis: yes, it is possible. This would require a download of the GraalVM distribution. I can make that happen.

File: src/main/scala/com/typesafe/sbt/packager/graalvm/GraalVMPlugin.scala (hunk 2 above)
Comment (ScalaWilliam) [reply src/main/scala/com/typesafe/sbt/packager/graalvm/GraalVMPlugin.scala]: I am not sure what you mean by "normalized".

File: src/main/scala/com/typesafe/sbt/packager/graalvm/Keys.scala (hunk 1 above)
Comment (ScalaWilliam) [reply src/main/scala/com/typesafe/sbt/packager/graalvm/Keys.scala]: Indeed a `native-image` specific setting.

**Comment by ScalaWilliam** [comment]
@muuki88 for some reason Travis is not even starting to run the tests any more (is it because I modified the travis yml?).

Can you take a look at the code again. I believe I improved the naming. I made the name of the plug-in longer, so we don't have conflict with any potential future GraalVM plugin (graalvm is much more than native-image).

**Comment by ScalaWilliam** [comment]
I wrote the Travis change to the best of my experience, but can't be sure it'll work until it runs ... :-)

**Comment by muuki88** [comment]
Let's give Travis some time. I found that sbt-native- packager often hits some rate limits.

I'll check as soon as I'm on my laptop of I find anything suspicious

**Comment by muuki88** [comment]
Meh. Really seems if there's a small error in the Travis file. It looks rather correct to me though 🤔

**Comment by ScalaWilliam** [comment]
Somehow, pushing it again reactivated Travis :-0

**Comment by muuki88** [comment]
Once again Travis needs some help 😣

File: .travis.yml [hunk 5]
```diff
@@ -36,6 +37,18 @@ before_install:
 
 before_script:
 - curl -L --create-dirs -o $HOME/.sbt/launchers/1.2.6/sbt-launch.jar http://dl.bintray.com/sbt/maven-releases/org/scala-sbt/sbt-launch/1.2.6/sbt-launch.jar
+# Make GraalVM `native-image` available and nothing else
+- if [[ ! -d graalvm-ce-1.0.0-rc7 ]]; then
+    cd $HOME;
+    wget -O graalvm-ce-1.0.0-rc7-linux-amd64.tar.gz https://github.com/oracle/graal/releases/download/vm-1.0.0-rc7/graalvm-ce-1.0.0-rc7-linux-amd64.tar.gz;
```
[the remark this answers is not in the document]
Comment (muuki88) [reply .travis.yml]: That's a very good question @huntc 🤔 thanks for bringing this up 🤗😀

I have no source to ask for legal advice like this. I'll try to do some research. Any hint is appreciated here 😘

[the remark this answers is not in the document]
Comment (muuki88) [reply .travis.yml]: The `community edition` is being used, which is at least free of fees. 

```
There is a community edition (CE) and an enterprise edition (EE) of GraalVM. The community edition is distributed under an open source license. It is free to use in production and comes with no strings attached, but also no guarantees or support
```
https://www.graalvm.org/docs/faq/

[the remark this answers is not in the document]
Comment (huntc) [reply .travis.yml]: Seems that license acceptance is required: https://www.oracle.com/technetwork/oracle-labs/program-languages/downloads/index.html

I’d consult the Travis team and/or request specific permission from Oracle.

[the remark this answers is not in the document]
Comment (muuki88) [reply .travis.yml]: @ScalaWilliams having technical and legals blockers here I vote for postponing the Travis integration into a separate pull request.

Sorry for the back and forth 😣

[the remark this answers is not in the document]
Comment (ScalaWilliam) [reply .travis.yml]: That's for Enterprise Edition, @huntc. Notice the "ee" in the `.gz` download name after you accept. Of course it's healthy to be wary of Oracle.

Community Edition is as follows:
https://github.com/oracle/graal#license
image

**Comment by ScalaWilliam** [comment]
I strengthened the test and removed Travis bits @muuki88. I've put the Travis bits in a separate branch here: https://github.com/ScalaWilliam/sbt-native-packager/tree/travis-graal

**Comment by ScalaWilliam** [comment]
Anything else before the merge? :-)

**Comment by muuki88** [comment]
Nope 😀

Was busy getting 30 today 🤣 I'll comment with a bit more details on how we could move forward regarding comments on twitter that a docker Integration would be nice. Will do so on Tuesday

**Comment by ScalaWilliam** [comment]
Regarding Docker integration: actually, this would apply to RPM as well as Universal as well as Debian.

In all these cases, we can package up the native binary and distribute it.

RPM/Debian + native:
- This is particularly interesting as normally for Java apps we require an external Java installation which is confusing to set up and maintain.
- Having them would be a HUGE win to get Scala to take a good healthy bite out of the Go/Rust pie.

Docker + native:
- Real use cases are for: A) final image size B) FaaS function-as-a-service = rapid start-up time. 
- But for services, where final image size is not a big deal, but high peak performance is, this is not so important.

**Comment by muuki88** [comment]
I'll try to release this tomorrow

**Comment by ScalaWilliam** [comment]
@muuki88 commit rights sound excellent! Thank you :-)

**Comment by muuki88** [comment]
Sweet :heart:  Welcome on board.

You can release simply by pushing a signed that with the version you want to release, e.g. `v1.3.12`

**Pull request merged**